### PR TITLE
REST: Create commit catalog handler fix

### PR DESCRIPTION
**Pull request by bryanck**
Currently in the REST catalog handler, during a CTAS, the files will be written and associated with the partition spec ID from the initial create request. When the commit occurs, an empty table metadata object is created and updates are applied to that. However, applying the updates will increment the partition spec ID, so the files will be associated with the empty spec ID, and thus partition pruning will not be available for these files.

This PR adds the ability to build on top of an empty TableMetadata so the initial IDs will match the initial files. It also moves `TableMetadata.newTableMetadata` to use the builder to instantiate a new instance so it can reuse functionality from the builder and to be consistent with `buildReplacement()`.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 1]
```diff
@@ -985,10 +1012,12 @@ public Builder addSnapshot(Snapshot snapshot) {
         return this;
       }
 
+      ValidationException.check(!schemas.isEmpty(), "Attempting to add a snapshot before a schema is added");
+      ValidationException.check(!specs.isEmpty(), "Attempting to add a snapshot before a partition spec is added");
+      ValidationException.check(!sortOrders.isEmpty(), "Attempting to add a snapshot before a sort order is added");
```
Comment (bryanck) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:1017]: When building metadata from empty, the table specs aren't initialized. These assertions ensure they are set before adding files, so files added during a transaction line up with the table specs.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 2]
```diff
@@ -985,10 +1012,12 @@ public Builder addSnapshot(Snapshot snapshot) {
...
       ValidationException.check(!snapshotsById.containsKey(snapshot.snapshotId()),
           "Snapshot already exists for id: %s",
           snapshot.snapshotId());
-
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:991]: Nit: unnecessary change that could cause commit conflicts.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 3]
```diff
@@ -1212,7 +1246,7 @@ public TableMetadata build() {
           ImmutableList.copyOf(newSnapshotLog),
           ImmutableList.copyOf(metadataHistory),
           ImmutableMap.copyOf(refs),
-          discardChanges ? ImmutableList.of() : ImmutableList.copyOf(changes)
+          discardChanges || base == null ? ImmutableList.of() : ImmutableList.copyOf(changes)
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:1249]: I'm not sure about this. If base is `null`, aren't all of the changes still valid?

Tests pass without this change, so I would probably not include it. What was the reason for adding this?

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 4]
```diff
@@ -1213,6 +1213,67 @@ public void testCompleteCreateTransaction() {
...
+
+    create.commitTransaction();
+
+    Assert.assertTrue("Table should exist after append commit", catalog.tableExists(TABLE));
+    Table table = catalog.loadTable(TABLE);
+
+    Assert.assertEquals("Table schema should match the new schema",
+        newSchema.asStruct(), table.schema().asStruct());
+    Assert.assertEquals("Table should have create partition spec", newSpec.fields(), table.spec().fields());
+    Assert.assertEquals("Table should have create sort order", newSortOrder.fields(), table.sortOrder().fields());
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:1267]: Minor: I think it would make more sense to describe the new schema, spec, and sort order as "updated" rather than "create" because these are actually different than the created ones.

Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:1267]: Also, do we need to validate the IDs of each of these?

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 5]
```diff
@@ -1213,6 +1213,67 @@ public void testCompleteCreateTransaction() {
...
+    Assert.assertEquals("Table properties should be a superset of the requested properties",
+        properties.entrySet(),
+        Sets.intersection(properties.entrySet(), table.properties().entrySet()));
+    if (!overridesRequestedLocation()) {
+      Assert.assertEquals("Table location should match requested", "file:/tmp/ns/table", table.location());
+    }
+    assertFiles(table, FILE_A, anotherFile);
+    assertFilePartitionSpec(table, FILE_A, 0);
+    assertFilePartitionSpec(table, anotherFile, 1);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:1276]: I think it would be better to use `TABLE_SPEC.specId()` and `newSpec.specId()` rather than hard-coding the IDs.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 3 above)
Comment (bryanck) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1249]: `TestLoadTableResponse.testHasOnlyKnownFields()` was failing, so my intent was to make it behave like `newTableMetadata` which was setting changes to empty, though probably my misunderstanding. I'll remove this and fix what's going on with that test.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 2 above)
Comment (bryanck) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:991]: OK I'll put that back, was making spacing consistent with the validation checks

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 3 above)
Comment (bryanck) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java]: I made this change

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 2 above)
Comment (bryanck) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java]: This is done

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java (hunk 5 above)
Comment (bryanck) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:1276]: The spec ID isn't incremented at this point, so `newSpec.specId()` is zero

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java (hunk 4 above)
Comment (bryanck) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java]: I updated the message to `updated`. and added ID validation.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java (hunk 5 above)
Comment (bryanck) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java]: I created constants for the expected IDs.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @bryanck! Looks great. And thanks for finding a better solution!
